### PR TITLE
Add feature to push measurements to S3

### PR DESCRIPTION
**Pull request by kamicut**
This commit allows the fetch process to push data measurements to S3
after adding them to the database.

The measurements are first added to the DB to determine if they're unique.
Unique records are then turned into CSV format and written to a user
specified AWS S3 bucket.

Pushing to S3 allows for:
1. Archiving the fetch data
2. Aggregation / Filtering in map-reduce pipelines using Amazon EMR or
Amazon Athena (could help with aggregations (#182))

### CODE REVIEW

**Comment by kamicut** [comment]
@jflasher note: I haven't added the raw data measurement in the CSV output and have only added potential aggregation columns. This would discard attribution and averagingPeriod tags. Do you think we should add those as columns?

**Comment by jflasher** [comment]
> I haven't added the raw data measurement in the CSV output and have only added potential aggregation columns. This would discard attribution and averagingPeriod tags. Do you think we should add those as columns?

I'd like S3 to become **the** authoritative source of the data (as opposed to the database). Because of this, somewhere in S3 we need to have all those fields. If the only downside is a bit more storage because we have all the data, let's just do that. 

We could also potentially keep the same format where we have major fields that we aggregate over and one JSON field. We will need to add fields in the future, so nice to have flexibility.

**Comment by kamicut** [comment]
@jflasher, ok, i'll add that. I've had trouble adding the JSON as a column because the CSV parser in Athena gets confused with the commas and double quotes of a JSON object. I think we would be able to properly parse the row with a specific delimiter like `|` or a regex.

File: fetch.js [hunk 1]
```diff
@@ -118,6 +117,43 @@ let buildSQLObject = function (m) {
...
+    latitude = m.coordinates.latitude || '';
+    longitude = m.coordinates.longitude || '';
+  }
+  row.push(latitude);
+  row.push(longitude);
+
+  // Add double quotes to each field
+  let quotedRow = row.map(item => `"${item}"`);
+
+  return quotedRow.join(',');
```
Comment (jflasher) [inline fetch.js]: Can we use something from http://csv.adaltas.com/ as we're already using `csv-parse` in the project? Think it'd be overall safer than making the CSV ourselves.

File: fetch.js [hunk 2]
```diff
@@ -283,6 +326,59 @@ const saveFetches = function (timeStarted, timeEnded, itemsInserted, err, result
 };
 
 /**
+ * Create a key using the current datetime
+ * @param {Date} d javascript Date() object
+ * @return {string} string in format yyyy-mm-dd/unixtime
+ */
+const getDateKey = function (d) {
```
Comment (jflasher) [inline fetch.js]: We've already got a dependency on `moment` so feel free to just use `const now = moment();` if you want.

File: fetch.js [hunk 3]
```diff
@@ -283,6 +326,59 @@ const saveFetches = function (timeStarted, timeEnded, itemsInserted, err, result
...
+    if (!validAWS) {
+      return done(new Error('missing AWS Credentials'));
+    }
+
+    log.info('Saving fetch to S3');
+
+    let s3 = new AWS.S3();
+
+    // Format into CSV
+    let rows = records.map(buildCSVObject).join('\n');
```
Comment (jflasher) [inline fetch.js]: Can maybe use http://csv.adaltas.com/ here as well?

**Comment by kamicut** [comment]
@jflasher I'm thinking of having fetch just post raw JSON measurements and scrap the CSV part. We can move the CSV forming to `bumblebee`. What do you think?

**Comment by jflasher** [comment]
Yeah, I think that'd be fine as long as the CSV was formatted to properly work with Athena (single line, etc). If we're transforming into something like Parquet for Athena, not an issue I guess.

**Comment by kamicut** [comment]
Yes, we would have a JSON bucket for the raw data (and it could have any number of fields). We then have a derived bucket that has CSV data that we can use for Athena aggregations or other downstream APIs. I have read about Parquet for Athena, but I think we would need another EC2 creating the files, because it requires a bit more processing that a lambda function can handle (if I'm not mistaken).

**Comment by jflasher** [comment]
> I have read about Parquet for Athena, but I think we would need another EC2 creating the files, because it requires a bit more processing that a lambda function can handle (if I'm not mistaken).

Yeah, I think this is true.

File: fetch.js [hunk 4]
```diff
@@ -11,7 +11,6 @@
 var argv = require('yargs')
   .usage('Usage: $0 --dryrun --source \'Beijing US Embassy\'')
   .boolean('dryrun')
-  .describe('dryrun', 'Run the fetch process but do not attempt to save to the database and instead print to console, useful for testing.')
```
Comment (jflasher) [inline fetch.js]: Did this get removed on purpose?

Comment (kamicut) [inline fetch.js]: Ah, no, that's weird. Good catch!

**Comment by jflasher** [comment]
This should be good to go, I'll test it out in the near future and look to get it merged in.

**Comment by jflasher** [comment]
Looked this over locally and it looked good so going to merge to staging then should go up to master.

**Pull request merged**